### PR TITLE
Prepare for Dart 3.0 api removals

### PR DESCRIPTION
**Pull request by jakemac53**
Fixes https://github.com/dart-lang/test/issues/1811

I also pre-emptively bumped our dep on matcher, which I think should be fine, but please push back if you disagree.

### CODE REVIEW

**Review by natebosch — APPROVED** [review]

File: legacy_tests/nnbd_opted_in_with_optout/mono_pkg.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# See https://pub.dev/packages/mono_repo
+
+sdk:
+- pubspec
+- stable # We will want to hardcode 2.19 once 3.x comes out
```
Comment (natebosch) [inline legacy_tests/nnbd_opted_in_with_optout/mono_pkg.yaml:5]: I'm not sure we need to run on stable at all. Running on 2.18 from the pubspec version should be enough.

**Review by natebosch — APPROVED** [review]

File: legacy_tests/nnbd_opted_in_with_optout/mono_pkg.yaml (hunk 1 above)
Comment (jakemac53) [reply legacy_tests/nnbd_opted_in_with_optout/mono_pkg.yaml]: SGTM, swapped all these to run on 2.18

**Comment by mit-mit** [comment]
Thanks Jake! Was this rolled?

**Comment by mit-mit** [comment]
Doesn't look like it was rolled; doing that in https://dart-review.googlesource.com/c/sdk/+/274542

**Pull request merged**